### PR TITLE
HA IPv6 support

### PR DESCRIPTION
**Pull request by martinfurmanski**

### CODE REVIEW

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java [hunk 1]
```diff
@@ -123,17 +125,26 @@ public URI switchToMaster( LifeSupport haCommunicationLife, URI me )
...
         String hostname = config.get( HaSettings.ha_server ).getHost();
-        if ( hostname == null || hostname.contains( "0.0.0.0" ) )
+        InetSocketAddress masterSocketAddress = masterServer.getSocketAddress();
+
+        if ( hostname == null || isWildcard( hostname ) )
         {
-            String masterAddress = ServerUtil.getHostString( masterServer.getSocketAddress() );
-            hostname = masterAddress.contains( "0.0.0.0" ) ? me.getHost() : masterAddress;
+            InetAddress masterAddress = masterSocketAddress.getAddress();
+            hostname = masterAddress.isAnyLocalAddress() ? me.getHost() : ServerUtil.getHostString( masterSocketAddress );
```
Comment (spacecowboy) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java:134]: Does `isAnyLocalAddress` cover the `0.0.0.0` case? The test below suggests that it does, but just want to make sure

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java [hunk 2]
```diff
@@ -334,11 +335,23 @@ private boolean catchUpWithMaster( UpdatePuller updatePuller ) throws IllegalArg
...
+            host = "[" + host + "]";
+        }
+
         InstanceId serverId = config.get( ClusterSettings.server_id );
-        String host = hostString.contains( HighAvailabilityModeSwitcher.INADDR_ANY ) ? me.getHost() : hostString;
-        return URI.create( "ha://" + host + ":" + port + "?serverId=" + serverId );
+        return URI.create( "ha://" + host + ":" + serverSocketAddress.getPort() + "?serverId=" + serverId );
+    }
+
+    private static boolean isWildcard( String hostString )
```
Comment (spacecowboy) [inline enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java]: Duplicate function, please de-duplicate :)

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    public void testClusterWithLocalhostAddresses() throws Throwable
+    {
+        GraphDatabaseService db = new HighlyAvailableGraphDatabaseFactory()
+                .newEmbeddedDatabaseBuilder( dir.makeGraphDbDir() )
+                .setConfig( ClusterSettings.cluster_server, "[::1]:5000" )
+                .setConfig( ClusterSettings.initial_hosts, "[::1]:5000" )
+                .setConfig( HaSettings.ha_server, "[::1]:6000" )
+                .setConfig( ClusterSettings.server_id, "1" )
+                .newGraphDatabase();
+
```
Comment (spacecowboy) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java]: I know you're just making sure there is no crash or such, but could you add something which actually uses the database? Create a node or something basic, just to really really verify that it works

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    public void testClusterWithWildcardAddresses() throws Throwable
+    {
+        GraphDatabaseService db = new HighlyAvailableGraphDatabaseFactory()
+                .newEmbeddedDatabaseBuilder( dir.makeGraphDbDir() )
+                .setConfig( ClusterSettings.cluster_server, "[::]:5000" )
+                .setConfig( ClusterSettings.initial_hosts, "[::1]:5000" )
+                .setConfig( HaSettings.ha_server, "[::]:6000" )
+                .setConfig( ClusterSettings.server_id, "1" )
+                .newGraphDatabase();
+
```
Comment (spacecowboy) [inline enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java]: Same comment as above

**Review by spacecowboy — APPROVED** [review]
Looks good so far.

Have you actually tested the changes by standing up a cluster in an IPv6(-only?) environment?

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java (hunk 2 above)
Comment (martinfurmanski) [reply enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToSlave.java]: It is actually subtly different. One checks a String straight out of config where I expect the [ ] format whereas the other one checks an address without the square brackets. I know it is crazy to have this code in this place this far down in the stack and duplicated, but I do not want to attempt a refactoring and cleanup.

File: enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java (hunk 1 above)
Comment (martinfurmanski) [reply enterprise/ha/src/main/java/org/neo4j/kernel/ha/cluster/SwitchToMaster.java:134]: I believe it does, yes.

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java (hunk 3 above)
Comment (martinfurmanski) [reply enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java]: Ok! I actually had that in place for my own sanity but removed it prior to pushing. I will add it back.

File: enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java (hunk 4 above)
Comment (martinfurmanski) [reply enterprise/ha/src/test/java/org/neo4j/kernel/ha/HaIPv6ConfigurationTest.java]: Same answer as above

**Comment by martinfurmanski** [comment]
@spacecowboy Not an IPv6-only environment, but local testing with IPv6 only configuration on my MBP, i.e. [::] and [::1] as ha.host.coordination and ha.host.data. All basic functionality seems to be working. System test pending.

**Comment by spacecowboy** [comment]
@martinfurmanski ok good. before merge I would like us to do a test using docker (or aws if you prefer that) to make sure we try it outside of a localhost environment

**Comment by martinfurmanski** [comment]
@spacecowboy That would be amazing, thanks!

**Review by digitalstain — APPROVED** [review]

**Comment by spacecowboy** [comment]
Running tests in a docker network, so far the database fails to bind to the ipv6 address. Will investigate why.

**Comment by spacecowboy** [comment]
Some testing:
IPV6 with globally routed address works
IPV6 with link-local address fails

@martinfurmanski already has a plan to fix the link-local problem (URI mangles the interface specification https://en.wikipedia.org/wiki/IPv6_address#Link-local_addresses_and_zone_indices)

**Comment by martinfurmanski** [comment]
Updated for link local address support in URIs.

**Review by spacecowboy — APPROVED** [review]
Tested and now it works with link-local address

**Pull request merged**